### PR TITLE
[WIP] C++17 Support

### PR DESCRIPTION
**Pull request by zpalmtree**
* Adds C++17 support
* Fixes a few compiler warnings caused by this
* Rewrites some of the p2p code sans boost to allow g++-7 to compile it
* Implements a few of the C++17 features to clean up the code and test it works as expected
* Extends travis builds to support gcc8, gcc7 on osx and linux, clang-6.0 and clang-5.0 on linux, and an unspecified clang on osx (apple has weird version numbers, and it's not trivial to install a specific clang version - i think we can change the clang version based on the osx-image though)

The p2p code could use some looking over. It was confusing as hell, I *think* I've ported it correctly, but not 100%. It finds peers, and syncs the blockchain correctly, at least.

I've updated the installation instructions to reflect the new packages needed, but have not checked they can be followed, yet.

Readme: https://github.com/turtlecoin/turtlecoin/blob/c%2B%2B17/README.md

### CODE REVIEW

**Comment by brandonlehmann** [comment]
aarch64-gcc-8 build via travis :)

uykgryk

**Comment by zpalmtree** [comment]
Well, all the checks have passed now. Just need to verify the [updated build instructions](https://github.com/turtlecoin/turtlecoin/blob/13d5128b0a99e047b8ae305c6e2bbb39cc511ff2/README.md) are correct, and a review on the p2p code changes would be appreciated.

**Comment by brandonlehmann** [comment]
Well... now that the build checks pass I'm willing to look at the code :) lol

File: external/rocksdb/CMakeLists.txt [hunk 1]
```diff
@@ -181,6 +181,12 @@ else()
   endif()
 endif()
 
+if(APPLE AND CMAKE_C_COMPILER_ID STREQUAL "GNU")
+    # Need to build against libc++ instead of libstc++ on apple
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -nostdinc++ -I/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1 -nodefaultlibs -lc++ -lc++abi -lm -lc -lgcc -lgcc_eh")
+endif()
+
+
```
Comment (brandonlehmann) [inline external/rocksdb/CMakeLists.txt]: What's this extra whitespace for?

File: README.md [hunk 2]
```diff
@@ -6,63 +6,93 @@
...
 #### Apple
 
 ##### Prerequisites
 
-- Install [cmake](https://cmake.org/). See [here](https://stackoverflow.com/questions/23849962/cmake-installer-for-mac-fails-to-create-usr-bin-symlinks) if you are unable call `cmake` from the terminal after installing.
+- Install [cmake](https://cmake.org/). See [here](https://stackoverflow.com/questions/23849962/cmake-installer-for-mac-fails-to-create-usr-bin-symlinks) if you are unable to call `cmake` from the terminal after installing.
 - Install the [boost](http://www.boost.org/) libraries. Either compile boost manually or run `brew install boost`.
 - Install XCode and Developer Tools.
 
+
```
Comment (brandonlehmann) [inline README.md]: Blank Line?

File: external/rocksdb/CMakeLists.txt [hunk 3]
```diff
@@ -251,7 +257,7 @@ if(ENABLE_THREAD_LOCAL)
   endif()
 endif()
 
-option(FAIL_ON_WARNINGS "Treat compile warnings as errors" ON)
+option(FAIL_ON_WARNINGS "Treat compile warnings as errors" OFF)
```
Comment (brandonlehmann) [inline external/rocksdb/CMakeLists.txt]: don't know how i missed that before. good catch!

File: external/rocksdb/util/transaction_test_util.cc [hunk 4]
```diff
@@ -127,8 +128,9 @@ bool RandomTransactionInserter::DoInsert(DB* db, Transaction* txn,
 
   std::vector<uint16_t> set_vec(num_sets_);
   std::iota(set_vec.begin(), set_vec.end(), static_cast<uint16_t>(0));
-  std::random_shuffle(set_vec.begin(), set_vec.end(),
-                      [&](uint64_t r) { return rand_->Uniform(r); });
+
```
Comment (brandonlehmann) [inline external/rocksdb/util/transaction_test_util.cc]: why the

File: external/rocksdb/util/transaction_test_util.cc [hunk 5]
```diff
@@ -127,8 +128,9 @@ bool RandomTransactionInserter::DoInsert(DB* db, Transaction* txn,
...
+  std::shuffle(set_vec.begin(), set_vec.end(), std::random_device{});
+
```
Comment (brandonlehmann) [inline external/rocksdb/util/transaction_test_util.cc]: white space

File: src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp [hunk 6]
```diff
@@ -36,7 +36,7 @@ bool post_notify(IP2pEndpoint& p2p, typename t_parametr::request& arg, const Cry
 }
 
 template<class t_parametr>
-void relay_post_notify(IP2pEndpoint& p2p, typename t_parametr::request& arg, const net_connection_id* excludeConnection = nullptr) {
+void relay_post_notify(IP2pEndpoint& p2p, typename t_parametr::request& arg, const boost::uuids::uuid* excludeConnection = nullptr) {
```
Comment (brandonlehmann) [inline src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp]: giphy

File: src/P2p/NetNode.cpp [hunk 7]
```diff
@@ -310,15 +312,15 @@ std::string print_peerlist_to_string(const std::list<PeerlistEntry>& pl) {
...
+  void NodeServer::for_each_connection(std::function<void(CryptoNoteConnectionContext&, uint64_t)> f)
   {
     for (auto& ctx : m_connections) {
       f(ctx.second, ctx.second.peerId);
     }
   }
 
   //-----------------------------------------------------------------------------------
-  void NodeServer::externalRelayNotifyToAll(int command, const BinaryArray& data_buff, const net_connection_id* excludeConnection) {
+  void NodeServer::externalRelayNotifyToAll(int command, const BinaryArray& data_buff, const boost::uuids::uuid* excludeConnection) {
```
Comment (brandonlehmann) [inline src/P2p/NetNode.cpp]: giphy

File: src/P2p/NetNode.cpp [hunk 8]
```diff
@@ -1051,8 +1053,8 @@ std::string print_peerlist_to_string(const std::list<PeerlistEntry>& pl) {
 
   //-----------------------------------------------------------------------------------
 
-  void NodeServer::relay_notify_to_all(int command, const BinaryArray& data_buff, const net_connection_id* excludeConnection) {
-    net_connection_id excludeId = excludeConnection ? *excludeConnection : boost::value_initialized<net_connection_id>();
+  void NodeServer::relay_notify_to_all(int command, const BinaryArray& data_buff, const boost::uuids::uuid* excludeConnection) {
```
Comment (brandonlehmann) [inline src/P2p/NetNode.cpp]: giphy

File: src/P2p/NetNode.h [hunk 9]
```diff
@@ -161,10 +162,10 @@ namespace CryptoNote
     void on_connection_close(P2pConnectionContext& context);
 
     //----------------- i_p2p_endpoint -------------------------------------------------------------
-    virtual void relay_notify_to_all(int command, const BinaryArray& data_buff, const net_connection_id* excludeConnection) override;
+    virtual void relay_notify_to_all(int command, const BinaryArray& data_buff, const boost::uuids::uuid* excludeConnection) override;
```
Comment (brandonlehmann) [inline src/P2p/NetNode.h]: Can't we find something better?

File: src/P2p/NetNode.h [hunk 10]
```diff
@@ -161,10 +162,10 @@ namespace CryptoNote
...
     virtual bool invoke_notify_to_peer(int command, const BinaryArray& req_buff, const CryptoNoteConnectionContext& context) override;
-    virtual void for_each_connection(std::function<void(CryptoNote::CryptoNoteConnectionContext&, PeerIdType)> f) override;
-    virtual void externalRelayNotifyToAll(int command, const BinaryArray& data_buff, const net_connection_id* excludeConnection) override;
+    virtual void for_each_connection(std::function<void(CryptoNote::CryptoNoteConnectionContext&, uint64_t)> f) override;
+    virtual void externalRelayNotifyToAll(int command, const BinaryArray& data_buff, const boost::uuids::uuid* excludeConnection) override;
```
Comment (brandonlehmann) [inline src/P2p/NetNode.h]: impossiblequeruloushapuku-small

File: src/P2p/NetNodeCommon.h [hunk 11]
```diff
@@ -5,6 +5,8 @@
 
 #pragma once
 
+#include <boost/uuid/uuid.hpp>
```
Comment (brandonlehmann) [inline src/P2p/NetNodeCommon.h]: giphy

File: src/P2p/P2pNodeConfig.h [hunk 12]
```diff
@@ -17,6 +17,7 @@
 
 #pragma once
 
+#include <boost/uuid/uuid.hpp>
```
Comment (brandonlehmann) [inline src/P2p/P2pNodeConfig.h]: giphy

File: src/P2p/P2pProtocolDefinitions.h [hunk 13]
```diff
@@ -7,6 +7,8 @@
 
 #include "P2pProtocolTypes.h"
 
+#include <boost/uuid/uuid.hpp>
```
Comment (brandonlehmann) [inline src/P2p/P2pProtocolDefinitions.h]: bitch-please-o-rly-oh-really-say-what-the-big-lebowski-what-ffs-for-fucks-sake-gif

File: src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp (hunk 6 above)
Comment (zpalmtree) [reply src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp]: I hate typedefs

File: src/P2p/NetNode.h (hunk 9 above)
Comment (zpalmtree) [reply src/P2p/NetNode.h]: We could hack together a 128bit type I guess

File: src/P2p/P2pProtocolDefinitions.h (hunk 13 above)
Comment (RocksteadyTC) [reply src/P2p/P2pProtocolDefinitions.h]: Just in case you didnt catch this one, this face means "Donny, you're out of your fucking element."

**Comment by RocksteadyTC** [comment]
10/10 best review ever

**Comment by zpalmtree** [comment]
@rashedmyt Confirms the windows build instructions work with boost 1.68.

Linux build instructions probably still need some more work.

`sudo apt-add-repository "deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-6.0 main"` should work to add the ppa on ubuntu 14.04, but is not working on 16.04.

**Comment by zpalmtree** [comment]
Build instructions have been tested with both clang and gcc on Ubuntu 14, 16, and 18.

**Comment by SoreGums** [comment]
a general query, does the version of clang/gcc matter when it comes to building a static binary and running on different base Linux versions?

I think the answer is no, just checking though. Thinking about how one might auto-build packages for different flavours of Linux.

Like there is a difference between Alpine's musl libc and nearly all over linux distros running on kernels comiled with glibc. so does building in Ubuntu 18.x with latest versions clang/gcc mean there is a difference when running on Ubuntu 12.x?

Cheers.

**Comment by brandonlehmann** [comment]
@SoreGums 

Although we build 'static' binaries we actually do not currently statically link libc in the Unbuntu built release binaries due to pthreads. 

```
# ldd TurtleCoind
        linux-vdso.so.1 =>  (0x00007ffe84121000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f2968b08000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f29687ff000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f29685e2000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f2968218000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f2968d10000)
```

Examining the aarch64 & osx builds; however, they do appear to be entirely static binaries.

That being said, I grabbed the latest release binaries and threw them on an old out of date CentOS 5 image the results were, of course, no bingo.

```
# ldd TurtleCoind
./TurtleCoind: /lib64/libpthread.so.0: version `GLIBC_2.12' not found (required by ./TurtleCoind)
./TurtleCoind: /lib64/libc.so.6: version `GLIBC_2.15' not found (required by ./TurtleCoind)
./TurtleCoind: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by ./TurtleCoind)
./TurtleCoind: /lib64/libc.so.6: version `GLIBC_2.8' not found (required by ./TurtleCoind)
./TurtleCoind: /lib64/libc.so.6: version `GLIBC_2.9' not found (required by ./TurtleCoind)
./TurtleCoind: /lib64/libc.so.6: version `GLIBC_2.7' not found (required by ./TurtleCoind)
        linux-vdso.so.1 =>  (0x00007fff38de4000)
        librt.so.1 => /lib64/librt.so.1 (0x0000003130600000)
        libm.so.6 => /lib64/libm.so.6 (0x000000312fe00000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x000000312fa00000)
        libc.so.6 => /lib64/libc.so.6 (0x000000312f200000)
        /lib64/ld-linux-x86-64.so.2 (0x000000312ee00000)
```

Throwing them on a more modern CentOS 7 image, resulted in the binaries working as expected as a compatible version of libc was located.

```
# ldd TurtleCoind
        linux-vdso.so.1 =>  (0x00007ffcee1dd000)
        librt.so.1 => /lib64/librt.so.1 (0x00007fa029ebb000)
        libm.so.6 => /lib64/libm.so.6 (0x00007fa029bb9000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fa02999d000)
        libc.so.6 => /lib64/libc.so.6 (0x00007fa0295d0000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fa02a0c3000)
```

**Comment by brandonlehmann** [comment]
#yolo

**Pull request merged**